### PR TITLE
Dq rules severity level

### PR DESCRIPTION
**Pull request by aysegulcayir**
-added get_highest_severity_from_validation_result function
-added severity field in rule class
-added this function documentation in readme part

### CODE REVIEW

File: src/dq_suite/common.py [hunk 1]
```diff
@@ -30,13 +31,18 @@ def __post_init__(self):
             if self.norm is not None:
                 raise TypeError("'norm' should be of type int")
 
+        if not isinstance(self.severity, str):
```
Comment (SSchotten) [inline src/dq_suite/common.py:34]: This one would then also be a Literal.

File: src/dq_suite/output_transformations.py [hunk 2]
```diff
@@ -571,3 +571,43 @@ def write_validation_result_tables(
...
+        "fatal": 3,
+        "error": 2,
+        "warning": 1,
+    }
+
+    for result in validation_result.get("results", []):
+        if result.get("success") is False:
+            expectation_type = result["expectation_config"]["type"]
+            expectation_type = expectation_type[len("expect_"):]
+            rule_name = "Expect" + "".join(
```
Comment (SSchotten) [inline src/dq_suite/output_transformations.py:602]: I don't really understand why this (+2 lines above) is necessary? As far as I know, all the rule names should be `WrittenLikeThis` in the input JSON.

File: src/dq_suite/common.py [hunk 3]
```diff
@@ -53,7 +59,7 @@ class RulesDict:
 
     unique_identifier: str  # TODO: List[str] for more complex keys?
     table_name: str
-    rules_list: RulesList
+    rules: RulesList
```
[the remark this answers is not in the document]
Comment (aysegulcayir) [reply src/dq_suite/common.py:62]: We are using this variable in input form as "rules", so I have changed it.

File: src/dq_suite/common.py [hunk 4]
```diff
@@ -30,13 +31,18 @@ def __post_init__(self):
             if self.norm is not None:
                 raise TypeError("'norm' should be of type int")
 
+        if (self.severity is not None and self.severity not in allowed_severities):
```
Comment (SSchotten) [inline src/dq_suite/common.py:34]: Is the value of `allowed_severities` defined?

File: src/dq_suite/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SSchotten) [reply src/dq_suite/common.py:62]: Ah, okay. I missed that, good point. 

In general though, they don't have to match up: if we can use a more descriptive name in the code that will help us interpret what's being processed, we are free to use it.

File: src/dq_suite/validation.py [hunk 5]
```diff
@@ -387,6 +390,10 @@ def run_validation(
         validation_settings_obj=validation_settings_obj,
     )
 
+    validation_result = list(checkpoint_result.run_results.values())[0]
+
+    highest_severity = get_highest_severity_from_validation_result(validation_result, rules_dict)
+
     if debug_mode:  # Don't write to UC in debug mode
         return checkpoint_result.success, checkpoint_result
```
Comment (SSchotten) [inline src/dq_suite/validation.py:398]: Should `highest_severity` maybe also be returned in debug mode?

File: src/dq_suite/output_transformations.py [hunk 6]
```diff
@@ -571,3 +572,37 @@ def write_validation_result_tables(
             table_name=validation_table_name,
             schema=schema,
         )
+
+
+def get_highest_severity_from_validation_result(validation_result: dict, rules_dict: dict) -> str | None:
```
Comment (SSchotten) [inline src/dq_suite/output_transformations.py:577]: I think the return type is now only string, because `None` is no longer possible.

File: docs/Readme-other.md [hunk 7]
```diff
@@ -41,3 +42,26 @@ The `dq_suite.validation.run_validation` function takes an `ms_teams_webhook` (s
...
+
+Severity levels (from highest to lowest):
+
+`fatal`: Fails the workflow. Workflow can't continue with this error.
+
+`error`: Does not fail the workflow, but should be fixed A.S.A.P.
+
+`warning`: Does not fail the workflow, but should be fixed eventually.
+
+If no failed expectations match any defined severity, the function returns None.
```
Comment (SSchotten) [inline docs/Readme-other.md:65]: Instead of `None`, it now returns "ok".

File: pyproject.toml [hunk 8]
```diff
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 
 [project]
 name = "dq-suite-amsterdam"
-version = "0.11.19"
+version = "0.11.20"
```
Comment (SSchotten) [inline pyproject.toml:7]: Could you also update `CHANGELOG.md` to reflect the changes in this version?

File: src/dq_suite/output_transformations.py (hunk 2 above)
Comment (SSchotten) [reply src/dq_suite/output_transformations.py]: Okay, understood. Interesting behaviour of GX then :) In that case using `humps` again seems to be the way indeed.

File: src/dq_suite/validation.py (hunk 5 above)
Comment (aysegulcayir) [reply src/dq_suite/validation.py:398]: I don't think it's necessary, at least not yet.

File: src/dq_suite/output_transformations.py [hunk 9]
```diff
@@ -270,6 +271,7 @@ def get_single_rule_dict(rule: Rule, table_id: str) -> dict:
         "norm": rule.get("norm", None),
         "bronTabelId": table_id,
         "attribuut": parameters.get("column", None),
+        "severity": rule.get("severity", None),
```
[the remark this answers is not in the document]
Comment (aysegulcayir) [reply src/dq_suite/output_transformations.py:274]: I will change as "ok".

File: src/dq_suite/validation.py (hunk 5 above)
Comment (SSchotten) [reply src/dq_suite/validation.py:398]: In that case, you could put the call to `get_highest_severity_from_validation_result` after the `if debug_mode` statement: because it's (currently) not relevant to debugging, so no need to calculate it when you activate debug mode.

File: tests/test_output_transformations.py [hunk 10]
```diff
@@ -450,6 +454,42 @@ def test_get_afwijking_data_raises_type_error(
...
+            },
+            {
+                "rule_name": "ExpectColumnValuesToBeUnique",
+                "parameters": {"column": "id"},
+                "severity": "fatal"
+            }
+        ]
+    }
+    result = get_highest_severity_from_validation_result(validation_result, rules_dict)
+    assert result == "fatal"
```
Comment (SSchotten) [inline tests/test_output_transformations.py:491]: I would also add a test to check that "ok" is returned when severity is absent in the `rules_dict` or when the results only contains `"success": True` values.

Comment (SSchotten) [reply tests/test_output_transformations.py:491]: Also, try adding a test case to `test_validation_input` catching the exception where `severity` is not equal to `None`, "fatal", "error" or "warning".

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,8 +4,15 @@ All notable changes to this project will be documented in this file.
...
+## [0.11.20] - 15-05-2025
+ 
+### Added
+
+### Changed
+- Added `highest severity level` to output
+- Modify `Rule` dataclass to include a severity field
 
-## [0.11.19] - 15-05-2025
+## [0.11.19] - 18-07-2025
```
[the remark this answers is not in the document]
Comment (aysegulcayir) [reply CHANGELOG.md:15]: Why did I do that? It was already 20 :D
